File: libs/python/mail_action_usps/vision.py
#!/usr/bin/env python3
"""
Vision analysis backends for USPS mailpiece scans.

Two backends:
  - openclaw_agent: copies image to agent workspace, asks openclaw agent to analyze
  - provided: analysis is passed directly (for Copilot inline / testing)

The backend is selected automatically: if analysis is supplied to the pipeline,
"provided" is used. Otherwise "openclaw_agent" is used.
"""

import json
import os
import re
import shutil
import subprocess
from pathlib import Path

# Load the analysis prompt from the sibling markdown file
_PROMPT_PATH = Path(__file__).parent / "analyze_prompt.md"
_ANALYSIS_PROMPT = ""
if _PROMPT_PATH.exists():
    _ANALYSIS_PROMPT = _PROMPT_PATH.read_text()


def _build_agent_prompt(staging_name: str) -> str:
    """Build the prompt sent to openclaw agent for a single image."""
    return (
        f"View the image at camera_captures/{staging_name} and analyze it.\n\n"
        f"{_ANALYSIS_PROMPT}\n\n"
        "Return ONLY the JSON object, no markdown fences, no explanation."
    )


def _get_agent_media_dir(agent: str) -> str:
    if not agent:
        raise ValueError("vision_agent is required")
    return os.path.expanduser(f"~/.openclaw/agents/{agent}/workspace/camera_captures")


def analyze_via_agent(image_path: str, vision_agent: str) -> dict:
    """Analyze a single mailpiece scan via openclaw agent vision."""
    agent_media_dir = _get_agent_media_dir(vision_agent)
    os.makedirs(agent_media_dir, exist_ok=True)

    src = Path(image_path)
    staging_name = f"usps-scan-{src.name}"
    staging_path = Path(agent_media_dir) / staging_name

    try:
        shutil.copy2(str(src), str(staging_path))
        prompt = _build_agent_prompt(staging_name)

        result = subprocess.run(
            [
                "openclaw", "agent", "--agent", vision_agent, "--json",
                "--timeout", "90", "--message", prompt,
            ],
            capture_output=True, text=True, timeout=120,
        )

        if result.returncode != 0:
            raise RuntimeError(f"openclaw agent failed: {result.stderr[:200]}")

        data = json.loads(result.stdout)
        text = data["result"]["payloads"][0]["text"]

        # Strip markdown fences if present
        text = re.sub(r"^```(?:json)?\s*", "", text.strip())
        text = re.sub(r"\s*```$", "", text.strip())
        return json.loads(text)

    finally:
        if staging_path.exists():
            staging_path.unlink()


def validate_analysis(analysis: dict) -> dict:
    """Ensure an analysis dict has all required fields with defaults."""
    return {
        "sender": analysis.get("sender", "Unknown"),
        "addressee": analysis.get("addressee", "Unknown"),
        "description": analysis.get("description", ""),
        "type": analysis.get("type", "scan"),
        "importance": analysis.get("importance", "medium"),
        "mail_class": analysis.get("mail_class", "Unknown"),
        "address_method": analysis.get("address_method", ""),
    }

File: libs/ts/mail_action_usps/src/vision.ts
/**
 * Vision analysis backends for USPS mailpiece scans.
 *
 * Two backends:
 *   - openclaw_agent: copies image to agent workspace, asks openclaw agent to analyze
 *   - provided: analysis is passed directly (for Copilot inline / testing)
 */

import {
  copyFileSync,
  existsSync,
  mkdirSync,
  readFileSync,
  unlinkSync,
} from "node:fs";
import { execFileSync } from "node:child_process";
import { basename, dirname, join } from "node:path";
import { homedir } from "node:os";

// Load the analysis prompt from the sibling markdown file
const PROMPT_PATH = join(dirname(new URL(import.meta.url).pathname), "analyze_prompt.md");
let ANALYSIS_PROMPT = "";
try {
  if (existsSync(PROMPT_PATH)) {
    ANALYSIS_PROMPT = readFileSync(PROMPT_PATH, "utf-8");
  }
} catch {
  // Prompt file not found — not critical
}

function buildAgentPrompt(stagingName: string): string {
  return (
    `View the image at camera_captures/${stagingName} and analyze it.\n\n` +
    `${ANALYSIS_PROMPT}\n\n` +
    "Return ONLY the JSON object, no markdown fences, no explanation."
  );
}

function getAgentMediaDir(agent: string): string {
  if (!agent) {
    throw new Error("vision_agent is required");
  }
  return join(
    homedir(),
    ".openclaw",
    "agents",
    agent,
    "workspace",
    "camera_captures",
  );
}

export interface AnalysisResult {
  sender: string;
  addressee: string;
  description: string;
  type: string;
  importance: string;
  mail_class: string;
  address_method: string;
}

/**
 * Analyze a single mailpiece scan via openclaw agent vision.
 */
export function analyzeViaAgent(
  imagePath: string,
  visionAgent: string,
): AnalysisResult {
  const agentMediaDir = getAgentMediaDir(visionAgent);
  mkdirSync(agentMediaDir, { recursive: true });

  const srcName = basename(imagePath);
  const stagingName = `usps-scan-${srcName}`;
  const stagingPath = join(agentMediaDir, stagingName);

  try {
    copyFileSync(imagePath, stagingPath);
    const prompt = buildAgentPrompt(stagingName);

    const stdout = execFileSync(
      "openclaw",
      [
        "agent",
        "--agent",
        visionAgent,
        "--json",
        "--timeout",
        "90",
        "--message",
        prompt,
      ],
      { timeout: 120000, encoding: "utf-8", stdio: ["pipe", "pipe", "pipe"] },
    );

    const data = JSON.parse(stdout);
    let text: string = data.result.payloads[0].text;

    // Strip markdown fences if present
    text = text.trim().replace(/^```(?:json)?\s*/, "");
    text = text.trim().replace(/\s*```$/, "");
    return JSON.parse(text);
  } finally {
    try {
      if (existsSync(stagingPath)) unlinkSync(stagingPath);
    } catch {
      // cleanup best-effort
    }
  }
}

/**
 * Ensure an analysis dict has all required fields with defaults.
 */
export function validateAnalysis(
  analysis: Record<string, unknown>,
): AnalysisResult {
  return {
    sender: (analysis.sender as string) ?? "Unknown",
    addressee: (analysis.addressee as string) ?? "Unknown",
    description: (analysis.description as string) ?? "",
    type: (analysis.type as string) ?? "scan",
    importance: (analysis.importance as string) ?? "medium",
    mail_class: (analysis.mail_class as string) ?? "Unknown",
    address_method: (analysis.address_method as string) ?? "",
  };
}
